### DISCUSSION TITLE
How to prevent Deno's fetch implementation from sending default headers?

### DISCUSSION THREAD

**Discussion by jsejcksn** [denoland/deno#9491] [Q&A]
I noticed that Deno adds default headers to every fetch request. They are even included as expected values in [this unit test](https://github.com/denoland/deno/blob/ca55a5d36efd947ba9e7c8e56baacfc8a0eba120/cli/tests/unit/fetch_test.ts#L640-L643).

Here is an example:

```json
{
  "accept": "*/*",
  "user-agent": "Deno/1.7.2",
  "accept-encoding": "gzip, br",
  "host": "localhost:8000"
}
```

How can I prevent this so that a fetch request can be sent with no header values?

**Comment by kitsonk** [comment]
There are certain headers that are required to be sent by the user agent and certain ones which cannot be controlled by the user. It is all rather extensive but documented in the standard: https://fetch.spec.whatwg.org/#terminology-headers

**Reply by jsejcksn** [reply]
Hi @kitsonk thanks for replying. Thanks for the link as well.

I just read through the entire section "2.2.2. Headers".

I did read the section about forbidden headers (controlled by the user agent), which I am familiar with, but I did not see anything mentioned about required headers. Can you elaborate on that?

**Reply by jsejcksn** [reply]
> I did not see anything mentioned about required headers. Can you elaborate on that?

@kitsonk I've searched more about this and still am not finding anything. Will you please provide more information?

**Reply by kitsonk** [reply]
> A forbidden header name is a header name that is a byte-case-insensitive match for one of
>
> `Accept-Charset`
> `Accept-Encoding`
> `Access-Control-Request-Headers`
> `Access-Control-Request-Method`
> `Connection`
> `Content-Length`
> `Cookie`
> `Cookie2`
> `Date`
> `DNT`
> `Expect`
> `Host`
> `Keep-Alive`
> `Origin`
> `Referer`
> `TE`
> `Trailer`
> `Transfer-Encoding`
> `Upgrade`
> `Via`
> or a header name that starts with a byte-case-insensitive match for `Proxy-` or `Sec-` (including being a byte-case-insensitive match for just `Proxy-` or `Sec-`).
> 
>  > These are forbidden so the user agent remains in full control over them. Names starting with `Sec-` are reserved to allow new headers to be minted that are safe from APIs using fetch that allow control over headers by developers, such as XMLHttpRequest. [XHR]

So this means that Deno get to choose to set those and that API consumers cannot control them.

[§ 4.8](https://fetch.spec.whatwg.org/#fetching) specifies that when `Accept` isn't set, it is to be set to `*/*` and [§ 4.6.11](https://fetch.spec.whatwg.org/#http-network-or-cache-fetch) indicates that if the user agent is not set that a default value is set.  So while you can change the values of those two headers, you cannot omit them.

**Reply by jsejcksn** [reply]
Ok, so if I want to send a network request without any headers using Deno, how would I do it? Does [`Deno.createHttpClient`](https://doc.deno.land/builtin/unstable#Deno.createHttpClient) allow me to do this?
